include/wil/network.h

```diff
@@ -0,0 +1,1338 @@
+        }
+        m_sockaddr.Ipv4.sin_port = original_port;
+    }
+
+    inline void socket_address::set_address(const IN_ADDR* addr) WI_NOEXCEPT
+    {
+        WI_ASSERT(family() == AF_INET);
+
+        const auto original_port{m_sockaddr.Ipv4.sin_port};
+        reset(AF_INET);
+        m_sockaddr.Ipv4.sin_addr.s_addr = addr->s_addr;
+        m_sockaddr.Ipv4.sin_port = original_port;
+    }
+
+    inline void socket_address::set_address(const IN6_ADDR* addr) WI_NOEXCEPT
+    {
+        WI_ASSERT(family() == AF_INET6);
+
+        const auto original_port{m_sockaddr.Ipv6.sin6_port};
+        reset(AF_INET6);
+        m_sockaddr.Ipv6.sin6_addr = *addr;
+        m_sockaddr.Ipv6.sin6_port = original_port;
+    }
+
+#if defined(WIL_ENABLE_EXCEPTIONS)
+    inline void socket_address::set_sockaddr(SOCKET s)
+    {
+        THROW_IF_FAILED(set_sockaddr_nothrow(s));
+    }
+
+    inline void socket_address::set_sockaddr(PCWSTR address)
+    {
+        THROW_IF_FAILED(set_sockaddr_nothrow(address));
+    }
+
+    inline void socket_address::set_sockaddr(PCSTR address)
+    {
+        THROW_IF_FAILED(set_sockaddr_nothrow(address));
+    }
+#endif
+
+    inline HRESULT socket_address::set_sockaddr_nothrow(SOCKET s) WI_NOEXCEPT
+    {
+        reset(AF_UNSPEC);
+
+        auto nameLength{length()};
+        auto error{::getsockname(s, sockaddr(), &nameLength)};
+        if (error != 0)
+        {
+            error = ::WSAGetLastError();
+            RETURN_WIN32(error);
+        }
+        return S_OK;
+    }
+
+    inline HRESULT socket_address::set_sockaddr_nothrow(PCWSTR address) WI_NOEXCEPT
+    {
+        PCWSTR terminator_unused;
+
+        reset(AF_INET);
+        constexpr BOOLEAN strict_string{TRUE};
+        if (RtlIpv4StringToAddressW(address, strict_string, &terminator_unused, in_addr()) == 0)
+        {
+            m_sockaddr.si_family = AF_INET;
+            return S_OK;
+        }
+
+        reset(AF_INET6);
+        if (RtlIpv6StringToAddressW(address, &terminator_unused, in6_addr()) == 0)
+        {
+            m_sockaddr.si_family = AF_INET6;
+            return S_OK;
+        }
+
+        reset(AF_UNSPEC);
+        return E_INVALIDARG;
+    }
+
+    inline HRESULT socket_address::set_sockaddr_nothrow(PCSTR address) WI_NOEXCEPT
+    {
+        PCSTR terminator_unused;
+
+        reset(AF_INET);
+        constexpr BOOLEAN strict_string{TRUE};
+        if (RtlIpv4StringToAddressA(address, strict_string, &terminator_unused, in_addr()) == 0)
+        {
+            m_sockaddr.si_family = AF_INET;
+            return S_OK;
+        }
+
+        reset(AF_INET6);
+        if (RtlIpv6StringToAddressA(address, &terminator_unused, in6_addr()) == 0)
+        {
+            m_sockaddr.si_family = AF_INET6;
+            return S_OK;
+        }
+
+        reset(AF_UNSPEC);
+        return E_INVALIDARG;
+    }
+
+    inline HRESULT socket_address::write_address_nothrow(socket_address_wstring& address) const WI_NOEXCEPT

```

Comment:
`write_*` suggests writing to a file or output stream. Should this be like "format"? Or "get"?

Reply:
ah... 'format'. that semantic works much better. thanks! I'll change this to format_address_*